src/contracts/atlas/AtlETH.sol

```diff
@@ -28,41 +28,99 @@ abstract contract AtlETH is Permit69 {
 
-    function balanceOf(address account) public view returns (uint256) {
-        return _escrowAccountData[account].balance;
+    function accountLastActiveBlock(address account) external view returns (uint256) {
+        return uint256(nonces[account].lastAccessed);
     }
 
-    function accountLastActiveBlock(address account) external view returns (uint256) {
-        return uint256(_escrowAccountData[account].lastAccessed);
+    function balanceOf(address account) public view returns (uint256) {
+        EscrowAccountData memory accountData = _balanceOf[account];
+        return uint256(accountData.balance - accountData.holds);
     }
 
     // Deposit ETH and get atlETH in return.
     function deposit() external payable {
-        _checkIfUnlocked();
+        // _checkIfUnlocked();
         _mint(msg.sender, msg.value);
     }
 
     // Redeem atlETH for ETH.
+    function redeem(uint256 amount) external {
+        _checkIfUnlocked();
+        _checkTransfersAllowed(msg.sender);
+
+        EscrowAccountData memory accountData = _balanceOf[msg.sender];
+
+        uint128 _amount = uint128(amount);
+
+        if (accountData.balance - accountData.holds < _amount) revert InsufficientBalance();
+
+        EscrowNonce memory nonceData = nonces[msg.sender];
+
+        nonceData.lastAccessed = uint64(block.number);
+        nonceData.withdrawalAmount += _amount;
+        accountData.holds += _amount;
+
+        nonces[msg.sender] = nonceData;
+        _balanceOf[msg.sender] = accountData;
+    }
+
     function withdraw(uint256 amount) external {
         _checkIfUnlocked();
         _checkTransfersAllowed(msg.sender);
 
-        if (_escrowAccountData[msg.sender].balance < amount) revert InsufficientBalance();
-        _burn(msg.sender, amount);
+        _withdrawAccounting(amount, msg.sender);
+
+        emit Transfer(msg.sender, address(0), amount);
         SafeTransferLib.safeTransferETH(msg.sender, amount);
     }
 
+    function withdrawSurcharge(uint256 amount) external {
+        _checkIfUnlocked();
+
+        if (surcharge < amount) revert InsufficientSurchargeBalance(surcharge, amount);
+
+        unchecked {
+            surcharge -= amount;
+        }
+
+        // NOTE: Surcharges are not deducted from totalSupply.
+        _balanceOf[address(0xa7145)].balance += uint128(amount);
+    }
+
+    function _withdrawAccounting(uint256 amount, address spender) internal {
+        EscrowNonce memory nonceData = nonces[spender];
+        EscrowAccountData memory accountData = _balanceOf[spender];
+
+        uint128 _amount = uint128(amount);
+
+        if (nonceData.withdrawalAmount < _amount) {
+            revert InsufficientRedeemedBalance(uint256(nonceData.withdrawalAmount), amount);
+        }
+        if (accountData.balance < _amount) revert InsufficientAvailableBalance(uint256(accountData.balance), amount);
+
+        nonceData.lastAccessed = uint64(block.number);
```

Comment:
By updating `lastAccessed` here, a user will have to wait before calling `redeem`, which then triggers another wait before calling `withdraw`, in the case that the user withdraws funds partially each time

I.e. redeem - [wait] - withdraw - [wait] - redeem - [wait] - withdraw

Can we remove the `lastAccessed` here? And perhaps rename it to `delayInitBlock`?

Then I think the flow would be:
 redeem - [wait] - withdraw - [NO WAIT]- redeem - [wait] - withdraw

Reply:
let's revisit the structure here -  i wonder if there's a simpler structure that we can do that won't be as disruptive to the erc20 mechanics.

let's discuss on the next call.